vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,494 @@
+				pool.mu.Unlock()
+				return
+			}
+
+			err = initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			if err != nil {
+				log.Logf(0, "failed to initPool: %v", err)
+			} else {
+				pool.mu.Unlock()
+				<-pool.proxy.onTerminated
+				pool.mu.Lock()
+			}
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
+		}
+	}()
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	if err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+	return pool, nil
+}
+
+func (pool *pool) Count() int {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	return pool.count
+}
+
+func (pool *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	pool.mu.Lock()
+	proxy := pool.proxy
+	pool.mu.Unlock()
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (pool *pool) Close() error {
+	// TODO: pool.proxy.Call("Close")
+	pool.cancelRun()
+	return nil
+}
+
+type ProxyApp struct {
+	*rpc.Client
+	onTerminated chan bool
+	subProcess   subProcessCmd // is needed for mocking
+	params       *proxyAppParams
+}
+
+type stdInOutCloser struct {
+	io.ReadCloser
+	io.WriteCloser
+}
+
+func (s stdInOutCloser) Close() error {
+	readErr := s.ReadCloser.Close()
+	writeErr := s.WriteCloser.Close()
+	if readErr != nil || writeErr != nil {
+		return fmt.Errorf("stdInOutCloser.Close() error on read: %v, write: %v", readErr, writeErr)
+	}
+	return nil
+}
+
+func runProxyApp(ctx context.Context, params *proxyAppParams, cmd string) (*ProxyApp, error) {
+	subProcess := params.CommandRunner(cmd)
+	freeOnError := []io.Closer{}
+	freeAll := func() {
+		for _, closer := range freeOnError {
+			closer.Close()
+		}
+	}
+
+	subStdout, err := subProcess.StdoutPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdoutpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subStdout)
+
+	subStdin, err := subProcess.StdinPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdinpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subStdin)
+
+	subprocessLogs, err := subProcess.StderrPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stderrpipe: %w", err)
+	}
+	freeOnError = append(freeOnError, subprocessLogs)
+
+	codec := jsonrpc.NewClientCodec(stdInOutCloser{
+		subStdout,
+		subStdin,
+	})
+
+	go io.Copy(os.Stdout, subprocessLogs)
+
+	err = subProcess.Start()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to start command %v: %w", cmd, err)
+	}
+
+	// TODO: check onTerminated is really needed
+	onTerminated := make(chan bool, 1)
+	waitDone := make(chan bool, 1)
+
+	go func() {
+		err := subProcess.Wait()
+		if err != nil {
+			log.Logf(0, "failed to Wait() subprocess: %v", err)
+		}
+		waitDone <- true
+	}()
+
+	go func() {
+		select {
+		case <-waitDone:
+		case <-ctx.Done():
```

Comment:
It's good to wait for the subprocess always before we start another one (if there is a brief period of time when the old process is not released all resources and we start another one, potentially it can lead to some resource errors, e.g. the next init will fail to open some file in exclusive mode):
```
	go func() {
		select {
		case <-waitDone:
		case <-ctx.Done():
			subProcess.Kill()
			<-waitDone
		}
		onTerminated <- true
		freeAll()
	}()
```

However, it seems that this ctx only complicates things by introducing an unnecessary additional level of logic. I think cancelRun can do something along the lines of:
```
func (pool *pool) cancelRun() {
	pool.proxy.subProcess.Kill()
	pool.proxy = nil
}
```
And then we don't need Context, CancelFunc and this goroutine.
We just have the one above that waits for the subprocess and sends on onTerminated channel.

Reply:
Redesigned. Use exec.CommandContext() now.